src/Mailer/TransportFactory.php

```diff
@@ -0,0 +1,114 @@
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         3.7.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Mailer;
+
+use Cake\Core\StaticConfigTrait;
+use InvalidArgumentException;
+
+/**
+ * Factory class for generating email transport instances.
+ */
+class TransportFactory
+{
+    use StaticConfigTrait;
+
+    /**
+     * Transport Registry used for creating and using transport instances.
+     *
+     * @var \Cake\Core\ObjectRegistry
+     */
+    protected static $_registry;
+
+    /**
+     * An array mapping url schemes to fully qualified Transport class names
+     *
+     * @var array
+     */
+    protected static $_dsnClassMap = [
+        'debug' => 'Cake\Mailer\Transport\DebugTransport',
+        'mail' => 'Cake\Mailer\Transport\MailTransport',
+        'smtp' => 'Cake\Mailer\Transport\SmtpTransport',
+    ];
+
+    /**
+     * Returns the Transport Registry used for creating and using transport instances.
+     *
+     * @return \Cake\Core\ObjectRegistry
+     */
+    public static function getRegistry()
+    {
+        if (!static::$_registry) {
+            static::$_registry = new TransportRegistry();
+        }
+
+        return static::$_registry;
+    }
+
+    /**
+     * Sets the Transport Registry instance used for creating and using transport instances.
+     *
+     * Also allows for injecting of a new registry instance.
+     *
+     * @param \Cake\Core\ObjectRegistry $registry Injectable registry object.
+     * @return void
+     */
+    public static function setRegistry(ObjectRegistry $registry)
+    {
+        static::$_registry = $registry;
+    }
+
+    /**
+     * Finds and builds the instance of the required tranport class.
+     *
+     * @param string $name Name of the config array that needs a tranport instance built
+     * @return void
+     * @throws \InvalidArgumentException When a tranport cannot be created.
+     */
+    protected static function _buildTransport($name)
+    {
+        if (!isset(static::$_config[$name])) {
+            throw new InvalidArgumentException(
+                sprintf('The "%s" transport configuration does not exist', $name)
+            );
+        }
+
+        if (is_array(static::$_config[$name]) && empty(static::$_config[$name]['className'])) {
+            throw new InvalidArgumentException(
+                sprintf('Transport config "%s" is invalid, the required `className` option is missing', $name)
+            );
+        }
+
+        static::getRegistry()->load($name, static::$_config[$name]);
```

Comment:
This method looks like what is sometimes found in the `_create` method of a registry implementation. What does having a factory + registry class provide us?

Reply:
What if we only had the registry? A registry in many ways is a memoized factory.

Reply:
If we only had the registry then we would have to make the registry use the `StaticConfigTrait` to provide the `setConfg()`, `getConfig()` etc. static methods.

Currently I have made the `TranportFactory`-`TranportRegistry` pair similar to `Cache`- `CacheRegistry` and `Log`-`LogEngineRegistry` pairs we have (without the extra methods `Cache` and `Log` have).

Reply:
Ok. Lets roll with that until the DI apocalypse arrives :smile:

Reply:
Lets keep this.